src/operators_dense.jl

```diff
@@ -234,6 +234,9 @@ permutesystems(a::AdjointOperator{B1,B2}, perm) where {B1<:CompositeBasis,B2<:Co
 
 identityoperator(::Type{S}, ::Type{T}, b1::Basis, b2::Basis) where {S<:DenseOpType,T<:Number} =
     Operator(b1, b2, Matrix{T}(I, length(b1), length(b2)))
+identityoperator(::Type{S}, ::Type{T}, b1::CompositeBasis, b2::CompositeBasis) where {S<:DenseOpType,T<:Number} =
+    Operator(b1, b2, kron([Matrix{T}(I, length(b1.bases[i]), length(b2.bases[i])) for i in reverse(1:length(b1.bases))]...))
+
```

Comment:
This is strange... I am probably missing something obvious, but it looks like these two variants should give exactly equal results for `CompositeBasis`?

Reply:
Yes, I agree they should, but from my testing they don't. Try running the following example:

> I don't know if this was intended but previously I found that:
> 
> ```julia
> b1a = GenericBasis(2)
> b1b = GenericBasis(3)
> b2a = GenericBasis(1)
> b2b = GenericBasis(4)
> b3a = GenericBasis(6)
> b3b = GenericBasis(5)
> I1 = dense(identityoperator(b1a, b1b))
> I2 = dense(identityoperator(b2a, b2b))
> I3 = dense(identityoperator(b3a, b3b))
> 
> a = tensor(I1,I3)
> a == identityoperator(a)
> ```
> 
> would result in false. To fix it, I implemented identityoperator() when having CompositeBasis. See the following for the implementation:
> 
> https://github.com/mabuni1998/QuantumOpticsBase.jl/blob/64da84f45beb5fc813b4cc5e4afb5ed7a6bf0fc8/src/operators_dense.jl#L237-L238
> 
> This was important because in LazyTensor we check whether the operator is identityoperator and when the above didn't return true it lead to unwanted errors.
> 
> https://github.com/mabuni1998/QuantumOpticsBase.jl/blob/64da84f45beb5fc813b4cc5e4afb5ed7a6bf0fc8/src/operators_lazytensor.jl#L115-L136

Reply:
That's odd. Could it be that the eltype is different? Maybe that `identityoperator()` method is broken?

Reply:
It's the data inside; they do not represent the same matrix. Maybe Julias UnitScaling follows another convention than we expect, or maybe the example I'm making shouldn't return true?

To make it even clearer, consider:

```julia
b1a = GenericBasis(2)
b1b = GenericBasis(3)
I1 = identityoperator(b1a, b1b)
I2 = identityoperator(b1b, b1a)
a = tensor(I1,I2)
julia> a.data
6×6 SparseArrays.SparseMatrixCSC{ComplexF64, Int64} with 4 stored entries:
 1.0+0.0im      ⋅          ⋅          ⋅          ⋅          ⋅
     ⋅      1.0+0.0im      ⋅          ⋅          ⋅          ⋅
     ⋅          ⋅          ⋅      1.0+0.0im      ⋅          ⋅
     ⋅          ⋅          ⋅          ⋅      1.0+0.0im      ⋅
     ⋅          ⋅          ⋅          ⋅          ⋅          ⋅
     ⋅          ⋅          ⋅          ⋅          ⋅          ⋅
julia> identityoperator(a).data
6×6 SparseArrays.SparseMatrixCSC{ComplexF64, Int64} with 6 stored entries:
 1.0+0.0im      ⋅          ⋅          ⋅          ⋅          ⋅
     ⋅      1.0+0.0im      ⋅          ⋅          ⋅          ⋅
     ⋅          ⋅      1.0+0.0im      ⋅          ⋅          ⋅
     ⋅          ⋅          ⋅      1.0+0.0im      ⋅          ⋅
     ⋅          ⋅          ⋅          ⋅      1.0+0.0im      ⋅
     ⋅          ⋅          ⋅          ⋅          ⋅      1.0+0.0im
```

Reply:
Oh... Yeah, I think in this case they actually *should not* be equal. It's kind of funny to even define `identityoperator()` in the rectangular case. Clearly it's not the identity!

Reply:
To summarize, the two methods of `identityoperator()` above are only equivalent in case the left and right composite bases both have the same length and have the same number of factors, and the dimensions of the factors match. Otherwise the first definition (the original one) is correct - it actually produces an identity operator int he square case - where the second is not. The second version also fails in case the left basis has more factors than the right.

Should we consider raising a warning if `identityoperator()` is used in rectangular cases?

Reply:
+1 for a warning

Reply:
I see good point. But then I would like some input on how we should treat LazyTensor with non-equivalent left and right basises. See below

Reply:
Okay, so that method here should be removed again.